server/rpm/pbench-server.spec.j2

```diff
@@ -59,7 +59,7 @@ mv %{buildroot}/%{installdir}/%{static}/package.json %{buildroot}/%{installdir}
 
 %post
 # Install python dependencies
-pip3 --no-cache-dir install --prefix=/%{installdir} -r /%{installdir}/requirements.txt > /%{installdir}/pip3-install.log
+pip3 --no-cache-dir install -r /%{installdir}/requirements.txt > /%{installdir}/pip3-install.log
```

Comment:
I wonder how this impacts containers (vis a vis Mr. Jenkins).

Reply:
`tox` does the install with a prefix anyway, in the temporary tox venv, so that at least should have no real impact. The problems all seem to stem from our friend the psycho cop.

Reply:
Until we change the way the server installs code, we have to install using `--prefix`.  This causes `pip3` to install Python3 modules in the RPM installed library paths, which breaks maintenance of the server.

For some reason we dropped the setting of the `PYTHONPATH` in the `pbench-trampoline` code.

Like the agent side, we need to set the `PYTHONPATH` to include the `/opt/pbench-server` installed Python3 modules.

Reply:
[I've un-resolved this conversation, in light of Peter's reply to it -- @portante, I think it's now on you to mark it resolved....]

Could we finesse this situation by using `--user` instead of `--prefix`?  That would prevent conflicts with the RPM-installed packages, and I believe that the user directory tree is implicitly added to the `PYTHONPATH`.  All we would have to do is to ensure that the installation is done by/for the user which is going to be running the tools.  (If necessary, perhaps we could use a `--prefix` which specifies the directory which `--user` uses.)  That might at least _reduce_ the problem-space....

Reply:
It's an interesting possibility; I believe installation is done under `root`, however, whereas we want to target `pbench` where the service and cron jobs will run. I'm not sure I feel entirely comfortable trying to emulate `pip --user`, although I suppose it's well documented and stable. I've never experimented with `--user` ... does `sys.path` really get implicitly adjusted?

(Well, yes, apparently, by implicit import of the `site` package, which can also be used to show the default `sys.path` a bit more straightforwardly than our experiments during the jam: `python3 -m site` will show all the standard places, including reporting that my per-user `~/.local` doesn't exist. So, indeed, perhaps `--prefix ~pbench/.local` would actually do the job?? That would implicitly take care of the cron and service files, and the "enable" script which runs as `pbench`... but maybe not other install/configure tasks from Ansible that run under root.)

Hmm.

Reply:
> I'm not sure I feel entirely comfortable trying to emulate `pip --user`

Yup, but when selecting the smaller insect (i.e., the lesser of weevils)....

Of course, perhaps we don't need to emulate it -- perhaps `su pbench pip --user` would do what we need...except for the "other install/configure tasks from Ansible`, but perhaps those could be run under `su` as well...or, maybe there is some way that Ansible can become the user that we need for certain tasks....

Reply:
I think I've got this working, although it's actually `su pbench -c "pip3 ..."` and even that has problems because the installation is done under `root`, and we're logging the pip results into `/opt/pbench-server/`, which we can't write as `pbench`. The `%post` actually fixes this by changing the whole tree to be owned by `pbench` at the end; however doing that earlier causes files owned by root to be laid down meaning we'd need to fix the permissions initially to write the log file and then later go ahead and sweep the directory tree. Instead, I tried running the `su pbench -c "..."` as a sub-shell with the root parent redirecting `stdout` ... and I'm not sure this is any klunkier than the alternatives although I'm not thrilled, either.

In any case, both the cron jobs and service, running as `pbench` implicitly pick up the user libraries, the `pbench-server-enable.yml` playbook already runs as `pbench`. I'm still doing some tests on the rest of the playbook (including the crontab generation, which relies on running a local Pbench server script), but I'm hopeful I'll have a new commit up today.

This all makes me wonder if the `%postun` ought to be removing the `pbench` user; but I think that's actually created by Ansible rather than by the rpm...

Reply:
Hey!  Look what I found:  there's an environment variable [`PYTHONUSERBASE`](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONUSERBASE)...which indicates where `--user` goes to...which means that you can probably skip the `su pbench`, set `PYTHONUSERBASE`, use `pip install --user`, and then let the `%post` fix everything up.  And, if you need to run any Python scripts as `root`, you can use `PYTHONUSERBASE` (instead of `PYTHONPATH`) to find our stuff.

Weird, wild, and wonderful....

Reply:
Yeah, I read that earlier; but I don't think there's anything "magical" about it, and you'd need to set that in order for `site` to find the relocated user site (so far as I was able to determine). Instead, I created a user configuration file causing the pbench user to automatically include `/opt/pbench-server/lib` in the PATH along with the local user site packages. This seems to work nicely, and means our Python interactive commands will run without any manual Python configuration environment variables.

Check out the new commit ...